File: java/496.next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        if (nums1 == null || nums2 == null) {
            return null;
        }
        Map<Integer, Integer> map = new HashMap<>();
        Deque<Integer> stack = new LinkedList<>();
        for (int i = nums2.length - 1; i >= 0; i--) {
            int num = nums2[i];
            while (!stack.isEmpty() && stack.peekFirst() < num) {
                stack.pollFirst();
            }
            map.put(num, stack.isEmpty() ? -1 : stack.peekFirst());
            stack.offerFirst(num);
        }
        int[] ans = new int[nums1.length];
        for (int i = 0; i < nums1.length; i++) {
            ans[i] = map.get(nums1[i]);
        }
        return ans;
    }
}

File: python3/496.next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        next_greaters = {}
        
        stack = []
        
        for index, num in enumerate(nums2):
            while stack and num > nums2[stack[-1]]:
                last_index = stack.pop()
                next_greaters[nums2[last_index]] = num
            
            stack.append(index)
            
            
        ans = []
        for num in nums1:
            ans.append(next_greaters[num] if num in next_greaters else -1)
            
        return ans
        
